src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+    operator ErroratedFuture<ValuesT...> () && {
+      using DestErrorator = \
+        typename ErroratedFuture<ValuesT...>::errorator_type;
+      using ThisErrorator = errorator<WrappedAllowedErrorsT...>;
+
+      static_assert(!DestErrorator::template is_less_errorated_v<ThisErrorator>,
+                    "conversion is possible to more-or-eq errorated future!");
+      return plainify(ErroratedFuture<ValuesT...>(std::move(*this)));
+    }
+
+    // initialize future as failed without throwing. `make_exception_future()`
+    // internally uses `std::make_exception_ptr()`. cppreference.com shouldn't
+    // be misinterpreted when it says:
+    //
+    //   "This is done as if executing the following code:
+    //     try {
+    //         throw e;
+    //     } catch(...) {
+    //         return std::current_exception();
+    //     }",
+    //
+    // the "as if" is absolutely crucial because modern GCCs employ optimized
+    // path for it. See:
+    //   * https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=cce8e59224e18858749a2324bce583bcfd160d6c,
+    //   * https://gcc.gnu.org/ml/gcc-patches/2016-08/msg00373.html.
+    //
+    // This behavior, combined with `__cxa_exception_type()` for inspecting
+    // exception's type, allows for throw/catch-free handling of stateless
+    // exceptions (which is fine for error codes). Stateful jumbos would be
+    // actually a bit harder as `_M_get()` is private, and thus rethrowing is
+    // necessary to get to the state inside. However, it's not unthinkable to
+    // see another extension bringing operator*() to the exception pointer...
+    //
+    // TODO: we don't really need to `make_exception_ptr` each time. It still
+    // allocates memory underneath while can be replaced with single instance
+    // per type created on start-up.
+    template <class ErrorT,
+              class DecayedT = std::decay_t<ErrorT>,
+              bool IsAllowed = (std::is_same_v<WrappedAllowedErrorsT,
+                                               DecayedT> || ...),
+              class = std::enable_if_t<IsAllowed>>
+    future(ErrorT&& e)
+      : base_t(
+          seastar::make_exception_future<ValuesT...>(
+            errorator<WrappedAllowedErrorsT...>::make_exception_ptr(e))) {
+      // this is `fold expression` of C++17. Because of the SFINAE above
+      // the assertion cannot fail. Keeping it just in the case.
+      static_assert((... || (std::is_same_v<DecayedT, WrappedAllowedErrorsT>)),
+                    "ErrorT is not enlisted in errorator");
+    }
+
+    template <class ValueFuncT, class ErrorVisitorT>
+    auto safe_then(ValueFuncT&& valfunc, ErrorVisitorT&& errfunc) {
+      static_assert((... && std::is_invocable_v<ErrorVisitorT,
+                                                WrappedAllowedErrorsT>),
+                    "provided Error Visitor is not exhaustive");
+
+      using ValueFuncResult = std::result_of_t<ValueFuncT(ValuesT&&...)>;
```

Comment:
might want to use `std::invoke_result_t<>`? as `std::result_of_t<>` is deprecated in C++17.

Reply:
Yeah, need to switch.

Reply:
Switched.